pkg/cri/resource-manager/policy/builtin/balloons/balloons-policy.go

```diff
@@ -0,0 +1,941 @@
+		if largestIndex == -1 || value > largestValue {
+			largestIndex = index
+			largestValue = value
+		}
+	}
+	return largestIndex, largestValue
+}
+
+// resetCpuConf resets CPU configurations globally. All balloons can
+// be ignored, their CPU configurations will be applied later.
+func (p *balloons) resetCpuConf() error {
+	// Usual inputs:
+	// - p.allowed (cpuset.CPUset): all CPUs available for this
+	//   policy.
+	//
+	// Other inputs, if needed:
+	// - p.reserved (cpuset.CPUset): CPUs of ReservedResources
+	//   (typically for kube-system containers).
+	//
+	// Note: p.useCpuConf(balloon) will be called before assigning
+	// containers on the balloon, including the reserved balloon.
+	log.Errorf("NOT IMPLEMENTED: resetCpuConf available: %s; reserved: %s", p.allowed, p.reserved)
+	return nil
+}
+
+// useCpuConf configures CPUs of a balloon.
+func (p *balloons) useCpuConf(bln *Balloon) error {
+	// Usual inputs:
+	// - CPUs that cpuallocator has reserved for this balloon:
+	//   bln.Cpus (cpuset.CPUSet).
+	// - User-defined CPU configuration for CPUs of balloon of this type:
+	//   bln.Def.CpuConf (now map[string]string).
+	// - Current configuration(?): feel free to add data
+	//   structure for this. For instance policy-global p.cpuConfs,
+	//   or balloon-local bln.cpuConfs.
+	//
+	// Other input examples, if needed:
+	// - Requested CPU resources by all containers in the balloon:
+	//   p.requestedMilliCpus(bln).
+	// - Free CPU resources in the balloon: p.freeMilliCpus(bln).
+	// - Number of assigned containers: bln.ContainerCount().
+	// - Container details: access p.cch with bln.ContainerIDs().
+	// - User-defined CPU AllocatorPriority: bln.Def.AllocatorPriority.
+	// - All existing balloon instances: p.balloons.
+	// - CPU configurations by user: bln.Def.CpuConf (for bln in p.balloons)
+	log.Errorf("NOT IMPLEMENTED: useCpuConf Cpus: %s; CpuConf: %s", bln.Cpus, bln.Def.CpuConf)
+	return nil
+}
+
+// forgetCpuConf is called when CPUs of a balloon are released from duty.
+func (p *balloons) forgetCpuConf(bln *Balloon) {
+	// Usual inputs: see useCpuConf
+	log.Errorf("NOT IMPLEMENTED: forgetCpuConf Cpus: %s; CpuConf: %s", bln.Cpus, bln.Def.CpuConf)
+}
+
+func (p *balloons) newBalloon(blnDef *BalloonDef, confCpus bool) (*Balloon, error) {
+	var cpus cpuset.CPUSet
+	var err error
+	blnIdx := len(p.balloonsByDef(blnDef))
+	// Allocate CPUs
+	if blnDef == p.reservedBalloonDef ||
+		(blnDef == p.defaultBalloonDef && blnDef.MinCpus == 0 && blnDef.MaxCpus == 0) {
+		// The reserved balloon uses ReservedResources CPUs.
+		// So does the default balloon unless its CPU counts are tweaked.
+		cpus = p.reserved
+	} else {
+		cpus, err = p.cpuAllocator.AllocateCpus(&p.freeCpus, blnDef.MinCpus, blnDef.AllocatorPriority)
+		if err != nil {
+			return nil, balloonsError("could not allocate %d MinCpus for balloon %s[%d]: %w", blnDef.MinCpus, blnDef.Name, blnIdx, err)
+		}
+	}
+	bln := &Balloon{
+		Def:      blnDef,
+		Instance: blnIdx,
```

Comment:
Hmm... is this correctly updated/adjusted in balloons when another, potentially not the last by index balloon deflates to 0 and gets removed ?

Reply:
Good catch, it was definitely a bug. Fixed.